### PR TITLE
Add confidence scores for output sequences

### PR DESCRIPTION
**Pull request by Peiyang-Song**
### This PR adds confidence scores to be output with each sequence (i.e., each tactic suggestion).

### CODE REVIEW

File: cpp/generator.cpp [hunk 1]
```diff
@@ -102,15 +101,16 @@ std::vector<int64_t> tokenize(const char *input) {
 }
 
 /* Simulated Byt5 detokenizer that reverses the effects of its tokenizer */
-std::string detokenize(const std::vector<int64_t> &tokens) {
```
Comment (yangky11) [inline cpp/generator.cpp:105]: Detokenization seems to have nothing to do with the scores? Why do we pass the scores into this function?

File: cpp/generator.cpp [hunk 2]
```diff
@@ -173,9 +174,11 @@ int64_t sample(const std::vector<float> &logits, double temperature) {
 
   int64_t sampled_token = distribution(gen);
   assert(0 <= sampled_token && sampled_token < VOCAB_SIZE);
-  return sampled_token;
+  double sampled_logit = logits[sampled_token];
+  return {sampled_token, sampled_logit};
```
Comment (yangky11) [inline cpp/generator.cpp:178]: Logits are different from log-likelihood. Logits are raw outputs of the neural network, and they can be arbitrary numbers. Log-likelihood must be negative. In PyTorch, you can calculate log-likelihood from logits using https://pytorch.org/docs/stable/generated/torch.nn.functional.log_softmax.html#torch.nn.functional.log_softmax. Here you have to implement it in C++. Note that doing it naively may have numerical stability issues, but there is a simple trick (check PyTorch's implementation or ask me).

File: cpp/generator.cpp (hunk 1 above)
Comment (Peiyang-Song) [reply cpp/generator.cpp:105]: Right, detokenization does not need the scores as input. Technically we can factor that out and only pass the tokens into it, let me change this.

File: cpp/generator.cpp (hunk 2 above)
Comment (Peiyang-Song) [reply cpp/generator.cpp:178]: Added translation trick and fixed! Thanks!

#### Current demo:
image

File: cpp/generator.cpp [hunk 3]
```diff
@@ -120,9 +119,9 @@ std::string detokenize(const std::vector<int64_t> &tokens) {
     wcstombs(buf, ws.c_str(), l + 1);
     std::string s(buf);
     delete[] buf;
-    return s;
+    return {s};
```
Comment (yangky11) [inline cpp/generator.cpp:122]: Why do we need the `{}`?

File: cpp/generator.cpp [hunk 4]
```diff
@@ -120,9 +119,9 @@ std::string detokenize(const std::vector<int64_t> &tokens) {
...
   } catch (std::range_error) {
-    return "";
+    return {""};
```
Comment (yangky11) [inline cpp/generator.cpp:124]: Same here

File: cpp/generator.cpp (hunk 3 above)
Comment (Peiyang-Song) [reply cpp/generator.cpp:122]: Oh that's incorrect. I once turned it to a pair and forgot to turn it back. Let me delete that.

File: cpp/generator.cpp (hunk 4 above)
Comment (Peiyang-Song) [reply cpp/generator.cpp:124]: Also for this

Comment (Peiyang-Song) [reply cpp/generator.cpp]: Both fixed! Thanks!
